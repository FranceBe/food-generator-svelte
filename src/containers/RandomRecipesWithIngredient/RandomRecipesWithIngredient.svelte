<script>
    import { writable } from 'svelte/store'
    import H2 from 'components/H2/H2.svelte'
    import Input from 'components/Input/Input.svelte'
    import ButtonBurgerShape from 'components/ButtonBurgerShape/ButtonBurgerShape.svelte'
    import RecipeCard from 'components/RecipeCard/RecipeCard.svelte'
    import { getRandomInt, getRandomNItems } from 'services/random.service'

    let recipes = []
    let ingredient = ''
    let noRecipeFound = false
    const listOfTotalRecipes = writable([])
    const inputIngredient = writable('')
    export const getRandomRecipes = n => {
        const randomPage = ingredient ? getRandomInt(10) : getRandomInt(50)
        const url = ingredient ?
                `api/puppy/?q=${ingredient}&p=${randomPage}` :
                `api/puppy/?p=${randomPage}`
        fetch(url)
                .then(res => res.json())
                .then(data => {
                    const listOfRecipes = data && data.results.length > 0 && getRandomNItems(data.results, n)
                    noRecipeFound = !listOfRecipes
                    listOfTotalRecipes.set(listOfRecipes)
                })
                .catch(err => {
                    throw Error(err)
                })
    }
    listOfTotalRecipes.subscribe(value => {
        recipes = value
    })
    const handleChange = event => {
        inputIngredient.set(event.target.value)
    }
    inputIngredient.subscribe(value => {
        ingredient = value
    })
</script>

<style>
    div {
        display: flex;
        flex-direction: row;
    }

    .global-container {
        display: flex;
        flex-direction: column;
        flex-wrap: wrap;
        justify-content: center;
        align-items: center;
        margin: 20px 0;
    }

    .not-found {
        margin: 10px;
        color: #EE6663;
    }
</style>

<Input label={'Une envie de...'} onChange={handleChange}/>
<H2 title="...Combien voulez vous de propositions ?"/>
<div>
    <ButtonBurgerShape className="with-marge" onClick={() => getRandomRecipes(1)} text="1"/>
    <ButtonBurgerShape className="with-marge" onClick={() => getRandomRecipes(5)} text="5"/>
    <ButtonBurgerShape onClick={() => getRandomRecipes(10)} text="10"/>
</div>
{#if recipes.length > 0}
    <div class="global-container">
      {#each recipes as recipe}
          <RecipeCard title={recipe.title} href={recipe.href} ingredients={recipe.ingredients} src={recipe.thumbnail}/>
      {/each}
    </div>
{/if}
{#if noRecipeFound}
    <div class="not-found">
        Désolé, nous n&apos;avons trouvé aucune recette qui correspond :&apos;(
    </div>
{/if}
