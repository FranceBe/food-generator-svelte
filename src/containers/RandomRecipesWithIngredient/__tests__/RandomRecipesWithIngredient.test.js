import { render, fireEvent, cleanup } from '@testing-library/svelte'
import RandomRecipesWithIngredient from
  'containers/RandomRecipesWithIngredient/RandomRecipesWithIngredient.svelte'

describe('RandomRecipesWithIngredient', () => {
  afterEach(() => {
    cleanup()
  })
  it('should match snapshot', () => {
    const { container } = render(RandomRecipesWithIngredient)
    expect(container).toMatchSnapshot()
  })
  it('should render 2 h2 and an input', () => {
    const { container, getAllByText } = render(RandomRecipesWithIngredient)
    expect(container.getElementsByTagName('h2')).toHaveLength(2)
    expect(container.getElementsByTagName('input')).toHaveLength(1)
    expect(getAllByText('Une envie de...').length).toBe(1)
    expect(getAllByText('...Combien voulez vous de propositions ?').length).toBe(1)
  })
  it('should render 4 div and 3 buttons', () => {
    const { container, getAllByText } = render(RandomRecipesWithIngredient)
    expect(container.getElementsByTagName('div')).toHaveLength(4)
    expect(container.getElementsByTagName('button')).toHaveLength(3)
    expect(getAllByText('1').length).toBe(1)
    expect(getAllByText('5').length).toBe(1)
    expect(getAllByText('10').length).toBe(1)
  })
  it('should call fetch when a button is clicked', async () => {
    const { getByText } = render(RandomRecipesWithIngredient)
    const button = getByText('1')

    await fireEvent.click(button)
    expect(fetch).toHaveBeenCalledTimes(1)
  })
})
