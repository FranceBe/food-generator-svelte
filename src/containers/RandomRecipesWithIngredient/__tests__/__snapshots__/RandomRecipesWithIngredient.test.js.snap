// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomRecipesWithIngredient should match snapshot 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <label>
        <h2>
          Une envie de...
        </h2>
         
        <input
          name="ingredients"
        />
      </label>
       
      <div
        class="group-container"
      >
        <h2>
          ...Combien voulez vous de propositions ?
        </h2>
         
        <div
          class="button-container"
        >
          <button
            class="with-marge"
          >
            <span>
              1
            </span>
          </button>
           
          <button
            class="with-marge"
          >
            <span>
              5
            </span>
          </button>
           
          <button
            class=""
          >
            <span>
              10
            </span>
          </button>
        </div>
      </div>
    </div>
     
     
    
  </div>
</body>
`;
